### PR TITLE
Dk refactor users, repos, and 2 factor auth

### PR DESCRIPTION
**Pull request by DavidKnott**
Creates services and POROs for both download and uploading files to aws.
Refactors the 2 factor auth to only work when the user clicks "forgot password" on the login page
Implements what we talked about for the home controller having both index and show actions to show repos
Fixed merge conflicts with Ethan's branch and with master
28 tests total, all passing
All functionality still works
@jmejia @cews7 @Laszlo-JFLMTCO @ethanbennett

### CODE REVIEW

**Comment by Laszlo-JFLMTCO** [comment]
Than you @DavidKnott @ethanbennett This PR is very important to establish a new baseline with significant amount of changes including feature development and refactor. I'm OK to merge to be able to evaluate in a deployed environment the overall behavior.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,13 +1,13 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
-
   helper_method :current_user, :current_repo
 
   def current_user
-    current_user ||= User.find(session[:user_id]) if session[:user_id]
+    #possibly add in default new user, look into (null user object)
```
Comment (ethanbennett) [inline app/controllers/application_controller.rb]: I think we should make a point to stop pushing comments.

**Review by ethanbennett — APPROVED** [review]
This all looks good! It seems like the merging all our stuff didn't cause too many problems.

**Comment by jmejia** [comment]
Any way we can keep future PRs smaller? On the job, it's more likely you'll get meaningful feedback.

**Comment by ethanbennett** [comment]
@jmejia Definitely--we were just talking about the same thing. We're playing catch up from having not merged enough in the early stages.

**Pull request merged**